api/src/main/java/org/apache/iceberg/actions/DropTable.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions;
+
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+
+/**
+ * An action that deletes data, manifest, manifest lists in a table.
+ * <p>
+ * Reuires query engine to distribute parts of the work.
+ */
+public interface DropTable extends Action<DropTable, DropTable.Result> {
+
+  /**
+   * Passes an alternative delete implementation that will be used for manifests and data files.
+   * <p>
+   *
+   * @param deleteFunc a function that will be called to delete manifests and data files
+   * @return this for method chaining
+   */
+  DropTable deleteWith(Consumer<String> deleteFunc);
```

Comment:
Since unlike expire snapshots, this doesn't have a single threaded predecessor, I think it may make sense to use a serializable consumer here (which we would have to add as an interface) so that the implementation could delete on executors as well. I wanted to do this on expire snapshots but I think i missed the boat on that.

Reply:
The only concern with distributed deletes is it might overwhelm the underlying filesystem depending on the number of files to be deleted.

Reply:
I think that is a fair concern, but I also dislike that we have to bring the whole file list back to the driver (even if we can do it one partition at a time). I'm willing to pass on this for now but I think in the future we'll want a SerializableConsumer class

Reply:
I think we can change it in the expire snapshots action now if we want as we are moving to v2 APIs. Whatever we decide to use there has to be consistent with this logic too.

I am worried a bit about sending too many requests at the same time but not needing to bring any data to the driver may be worth it. Maybe, there can be a solution in between. Let me think about this for a while.

Reply:
cc @rdblue @shardulm94 @jackye1995 @yyanyy @prodeezy

Reply:
I think I am convinced now that doing this without bringing data to the driver would be reasonable. Also, we will have something in our Spark UI instead of having nothing and guessing if the job hangs or does something on the driver.

Reply:
> The only concern with distributed deletes is it might overwhelm the underlying filesystem depending on the number of files to be deleted.

I think we need to deal with this anyway through some throttling mechanism in the file system client. In the situation where there are many more data files than manifest files, and data files are using `ObjectStorageLoactionProvider`, distributed delete in the executor will be very beneficial for cloud storages.

Reply:
> I think we need to deal with this anyway through some throttling mechanism in the file system client

That's true but throttling is usually per JVM so we have to be careful with the number of machines we are going to use for deletes. Using `ObjectStoreLocationProvider` would help to distribute the load on the storage side.

It sounds like we need to come up with a generic way to represent such deletes, a reasonable default behavior, update v2 expire actions API and update our procedures to handle this.

Reply:
Do we need to control this beyond controlling the number of shuffle partitions? I'd say yes.

Reply:
Well the other approach is what we considered before, having a method which returns a dataframe of files to delete and leave it up to the user to decide the most efficient way to delete them if they want to do a distributed delete. This is much less user friendly but I know some folks are already using this approach with ExpireSnapshots.

Reply:
Discussed this with @aokolnychyi offline, came up with following proposal:
Introduce an config for distributed deletes(say `distributed_deletes`) and another config to control parallelism(say `delete_parallelism`)
1. `distributed_deletes` disabled
	1. `stream_results` enabled -  files will be deleted in a serial manner in the driver(Note: we pull the data on demand)
	2. `stream_results` disabled - file names are collected to the driver. `delete_parallelism` will be used to control the number of threads in the executor service, to perform delete.
2. `distributed_deletes` enabled
	`delete_parallelism` will be used used to coalesce the partitions and deletes performed in executors. When coalescing could result in overwhelming the executor memory, we could go with option **1.a**, where data is pulled to driver on demand.

Reply:
I'd like to hear from @prodeezy @fbocse. You guys contributed a way to pass an executor service to parallelize deletes. Is there a good use case where we need to pass an executor service directly or just controlling the parallelism is sufficient? As @karuppayya said, we could offer only an option to control the parallelism and use that to do deletes from executors without bringing data to the driver. For deletes on the driver, we can instantiate a new thread pool.

Reply:
I'm not sure it makes sense to parallelize deletes in Spark instead of in a thread-pool. What we've found is that this kind of thing ends up causing a high rate of transactions per second, which ends up getting throttled by S3 (in our case). Using a thread pool in the driver is a good middle ground because you get the parallelism, but don't waste cluster resources (task thread mostly idle while waiting on S3) and don't get throttled.

That said, if there are other object stores where that's not a problem, then maybe this makes more sense.

Reply:
I agree about the parallelism in general. Whether we do this on the driver or on executors, I think we will be limited by the underlying storage in most cases (maybe not all?). One of the points that @RussellSpitzer brought is the amount of data we will have to bring to the driver. By issuing deletes from executors after coalescing the partitions, we can avoid bringing data to the driver. Technically, `toLocalIterator` helps us to reduce the memory pressure but that's not ideal. The worry with coalescing partitions to have reasonable parallelism is that it may produce too large partitions anyway.

Reply:
@prodeezy @fbocse, which storage do you use? What's the max parallelism you have now?

Reply:
My worry with the distributed deletes is that we will need to coalesce all records into a smaller number of partitions, potentially having too large Spark tasks. If we have too many files to clean, streaming the results one partition at a time may be more scalable. I think it is okay to do whatever we do today with stream-results and using a thread pool on the driver. We can reconsider it later.

Reply:
Resolving this thread for now and keeping the behavior we have in other actions.